python/grass/grassdb/manage.py

```diff
@@ -50,6 +51,122 @@ def rename_location(database, old_name, new_name):
     grassdb, location = os.path.split(path)
     return grassdb, location, mapset
+
+
+class MapsetPath:
+    """This is a representation of a path to mapset.
+
+    Individual components are accessible through read-only properties.
+    It has with os.PathLike interface.
+
+    Paths are currently stored as is (not resolved, not expanded),
+    but that may change in the future.
+    """
+
+    def __init__(self, path, directory, location, mapset):
+        # Path as an attribute. Inheriting from Path would be something to consider
+        # here, however the Path inheritance is somewhat complex at this point.
+        self._path = Path(path)
+        self._directory = str(directory)
+        self._location = location
+        self._mapset = mapset
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}("
+            f"{self._path!r}, "
+            f"{self._directory!r}, {self._location!r}, {self._mapset!r})"
+        )
+
+    def __str__(self):
+        return str(self._path)
+
+    def __fspath__(self):
+        return os.fspath(self._path)
+
+    @property
+    def path(self):
+        """Full path to the mapset as a pathlib.Path object"""
+        return self._path
+
+    @property
+    def directory(self):
+        """Location name"""
+        return self._directory
+
+    @property
+    def location(self):
+        """Location name"""
+        return self._location
+
+    @property
+    def mapset(self):
+        """Mapset name"""
+        return self._mapset
+
+
+def resolve_mapset_path(path, location=None, mapset=None):
+    """Resolve full path to mapset from given combination of parameters.
+
+    Full or relative path to mapset can be provided as *path*. If the *path*
+    points to a valid location instead of a valid mapset, the mapset defaults
+    to PERMANENT.
+
+    Alternatively, location and mapset can be provided separately. In that case,
+    location and mapset are added to the path. If location is provided and mapset
+    is not, mapset defaults to PERMANENT.
+
+    The function does not enforce the existence of the directory or that it
+    it a mapset. It only manipulates the paths except for internal checks
+    to determine default values in some cases.
+
+    Home represented by ``~`` (tilde) and relative paths are resolved
+    and the result contains absolute paths.
+
+    Returns object with attributes path, db, location, and mapset.
+    The path attribute is a libpath.Path object representing full path
+    to the mapset. The db attribute is full path to the directory where
+    the location directory is. The attributes location and mapset are names
+    of location and mapset, respectively.
+    """
+    # We reduce the top-level imports because this is initialization code.
+    # pylint: disable=import-outside-toplevel
+
+    # This also resolves symlinks which may or may not be desired.
+    path = Path(path).expanduser().resolve()
+    default_mapset = "PERMANENT"
```

Comment:
I would suggest to use permanent_mapset from config.py, just to have less hardcoded values.

Reply:
Makes sense. Done in 9a156b0. We need to improve names of variables in that file to make it more useful, BTW.